### PR TITLE
feat(PresentationalNumberInput): introduction

### PR DESCRIPTION
**Pull request by asudoh**
`<PresentationalNumberInput>` is presentational-only version of `<NumberInput>` to let application handle its state rather than handling it internally.

Refs carbon-design-system/carbon-components-react#2207.

#### Changelog

**New**

- `<PresentationalNumberInput>`, presentational-only version of `<NumberInput>` to let application handle its state rather than handling it internally.

**Changed**

- `<NumberInput>` to use `<PresentationalNumberInput>` internally.

#### Testing / Reviewing

Testing should make sure our `<NumberInput>` is not broken.

### CODE REVIEW

**Review by emyarod — APPROVED** [review]
looks good to me!

File: packages/react/src/components/NumberInput/NumberInput.js [hunk 1]
```diff
@@ -381,6 +236,320 @@ class NumberInput extends Component {
...
+    evt.imaginaryTarget = inputRef;
+    this.setState(
+      {
+        value,
+      },
+      () => {
+        const { disabled, onClick, onChange } = this.props;
+        if (disabled) {
+          // `<PresentationalNumberInput>` takes care of preventing event being fired in disabled state.
+          // The code here is for simiulated testing
```
Comment (emyarod) [inline packages/react/src/components/NumberInput/NumberInput.js]:
```suggestion
          // The code here is for simulated testing
```
just a small typo

**Review by joshblack — COMMENTED** [review]
`PresentationalNumberInput` feels verbose to type out, I would think for controlled/uncontrolled we should try and keep the same component and change depending on stuff passed in (similar to libs like downshift)

**Comment by asudoh** [comment]
Leaving a note; The notion of switching controlled/uncontrolled behavior upon props has been considered, stemming from easier implementation. The notion of separate component came along as a alternative for putting precedence on separation-of-concern and clear differentiation in behavior. The controlled version can be accessed in a similar manner as filtered-version of our multi-select component.

The argument wrt verbose name may still be fair enough. Depending on users' feedback we can consider flipping the switch.

**Comment by tw15egan** [comment]
What are some possible alternative component names we could use?

**Comment by asudoh** [comment]
@tw15egan Thank you for asking! The current choice is `<NumberInput.Presentational>` but we can do `<NumberInput.V>` (from "view"), etc.

**Comment by shixiedesign** [comment]
Is this something like a `read-only` number input? We have on-going work for `read-only` text input. See https://github.com/carbon-design-system/carbon/issues/2177

Another question, is there an issue related to this with more use case explanations? Thanks!

**Comment by asudoh** [comment]
@shixiedesign "View" is as-in traditional MVC (technical pattern). Sorry for the confusion 🙇

**Comment by asudoh** [comment]
Created Redux Form example as discussed internally: https://codesandbox.io/s/github/asudoh/carbon-components/tree/redux-form-example/packages/react/examples/redux-form

https://github.com/asudoh/carbon-components/blob/a66a284/packages/react/examples/redux-form/src/FieldLevelValidationForm.js#L25-L43 shows the gist of it.

**Comment by emyarod** [comment]
@asudoh should I rename and also create an example in #2655?

**Comment by joshblack** [comment]
I'd be a little worried about introducing the view nomenclature here, particularly because in the traditional MVC sense the view most often represents a model and in our case we're just talking about stateless render method/functions.

I think if we back up to first principles here, and take a look at where ideas of controlled/uncontrolled originate from, we can look at the Forms documentation on the React website: https://reactjs.org/docs/forms.html

In these cases, the component itself is `input`. From their example:

```jsx
<form>
  <label>
    Name:
    <input type="text" name="name" />
  </label>
  <input type="submit" value="Submit" />
</form>
```

> This form has the default HTML form behavior of browsing to a new page when the user submits the form. **If you want this behavior in React, it just works.**

The way we dip into controlled mode is by passing in `value` and `onChange` props. For example:

```jsx
  render() {
    return (
      <form onSubmit={this.handleSubmit}>
        <label>
          Name:
          <input type="text" value={this.state.value} onChange={this.handleChange} />
        </label>
        <input type="submit" value="Submit" />
      </form>
    );
  }
```

Our component stays the same (`<input>`) but now we are able to control internal state.

I think this would be a good thing to keep in mind with these decisions, and the closer we can get to this behavior the better 👍 If the goal here is to pull out stateful logic, then this would be a great opportunity to look at hooks as a natural way of separating out this logic to be re-used between components.

**Comment by vpicone** [comment]
Definitely agree, we need to be extremely consistent in our extraction, hooks are a great opportunity to share that logic.

**Comment by joshblack** [comment]
Interesting pattern from Dan that is basically what we were talking about yesterday in our Slack thread: https://twitter.com/dan_abramov/status/1065354652697444353

File: packages/react/src/components/NumberInput/NumberInput.js [hunk 2]
```diff
@@ -26,6 +26,345 @@ const defaultTranslations = {
...
+};
+
+NumberInputView.defaultProps = {
+  disabled: false,
+  hideLabel: false,
+  iconDescription: 'choose a number',
+  label: ' ',
+  onChange: () => {},
+  onClick: () => {},
+  step: 1,
```
Comment (vpicone) [inline packages/react/src/components/NumberInput/NumberInput.js:357]: Is the `step` prop here relevant?

File: packages/react/src/components/NumberInput/NumberInput.js [hunk 3]
```diff
@@ -26,6 +26,345 @@ const defaultTranslations = {
...
+  value: 0,
+  invalid: false,
+  invalidText: 'Provide invalidText',
+  ariaLabel: 'Numeric input field with increment and decrement buttons',
+  helperText: '',
+  light: false,
+  allowEmpty: false,
```
Comment (vpicone) [inline packages/react/src/components/NumberInput/NumberInput.js:364]: Should this be handled with `invalid`?

File: packages/react/src/components/NumberInput/NumberInput.js [hunk 4]
```diff
@@ -26,6 +26,345 @@ const defaultTranslations = {
...
+  label: ' ',
+  onChange: () => {},
+  onClick: () => {},
+  step: 1,
+  value: 0,
+  invalid: false,
+  invalidText: 'Provide invalidText',
+  ariaLabel: 'Numeric input field with increment and decrement buttons',
+  helperText: '',
+  light: false,
```
Comment (vpicone) [inline packages/react/src/components/NumberInput/NumberInput.js:363]: As long as we're making a new API can we switch to a `variant` approach and away from boolean props? Something like `variant="light"` to allow for future variations with props that don't conflict.

File: packages/react/src/components/NumberInput/index.js [hunk 5]
```diff
@@ -5,5 +5,10 @@
  * LICENSE file in the root directory of this source tree.
  */
 
+import NumberInput, { NumberInputView } from './NumberInput';
+
+NumberInput.View = NumberInputView;
```
Comment (vpicone) [inline packages/react/src/components/NumberInput/index.js:10]: While there's no inherent problem in also tacking it on here, I think we need to pick a usage method for our documentation and stick with it.

File: packages/react/src/components/NumberInput/NumberInput.js (hunk 2 above)
Comment (asudoh) [reply packages/react/src/components/NumberInput/NumberInput.js:357]: AFAIK it's for determining how many number increases by up/down arrow

**Comment by vpicone** [comment]
As far as naming is concerned, I think we can convey the distinction between the components better by sticking to react jargon.

maybe `NumberInput.Controlled = ControlledNumberInput`

**Comment by vpicone** [comment]
> Interesting pattern from Dan that is basically what we were talking about yesterday in our Slack thread: [twitter.com/dan_abramov/status/1065354652697444353](https://twitter.com/dan_abramov/status/1065354652697444353)

@asudoh this is looks like a really powerful pattern worth considering

File: packages/react/src/components/NumberInput/NumberInput.js (hunk 3 above)
Comment (asudoh) [reply packages/react/src/components/NumberInput/NumberInput.js:364]: Seems to have come from a contributor (#1352). `invalid` explicitly makes the UI show the invalid state, whereas `allowEmpty` does internal validation for user-entered empty string.

**Comment by asudoh** [comment]
While I wonder why Redux wasn't brought up in the tweet, @joshblack's point wrt following how React handles native `<input>` sounds fair enough. Let me close this PR, and will come up with a newer PR with "auto-switching" approach.

**Pull request closed without merging**